### DISCUSSION TITLE
How to configure for multi-module project?

### DISCUSSION THREAD

**Discussion by gotson** [jreleaser/jreleaser#1237] [Q&A]
Hello,

let's say i have a project that has multiple modules, for example:
- a library that i want to publish to Maven Central
- a CLI application that i want to publish to various packages like Docker and Homebrew

The version number would be shared between all modules, and if a release happens, a single release note would be created, listing both the library and CLI features/fixes, as well as artifacts.

How would i go about configuring JReleaser, assuming i would use the Gradle plugin?

Can it be configured at the root of the repo for the common parts to both modules, and with some additional configuration in each module to describe the distributions and packagers ?

**Comment by aalmiray** [comment]
JReleaser's configuration is currently not designed to be broken down into smaller chunks that can be assembled at a single location (root is leader, children provides what's specific to them), nor define a base configuration that may be pulled into or reused by a leaf project (root defines common configuration, children add their specifics). 

As such my recommendation is to configure JReleaser at a single module. This would typically be the root but it could by a child as well (make sure to set `gitRootSearch` otherwise it will fail).

**Reply by gotson** [reply]
I would foresee some issues when trying to configure the distributions for the subprojects, for example when needing to point to the right files produced by the build.

If defined at the root level, then string values must be used, which would be brittle.

For instance if configuring with Gradle, it would make sense to be able to tap into the gradle build model, but that can't be done from the top-level project unfortunately.

**Reply by aalmiray** [reply]
That's right. Gradle's preferred choice of using the Property/Provider API for defining lazy values won't work in this situation. You'd have to rely on plain values. I think that at least the following would have to happen to realize what you envision:

- create a `jreleaser-base` plugin that provides a `jreleaser` extension only. No task registration.
- update the `jreleaser` plugin to consume the configuration from the current project and potentially merge it with the configuration from the project's parent (this is a tricky thing to do thanks to timings and the often shunned `afterEvaluate` block).

This _could_ solve the configuration issue but then what about the execution model? A release should happen as a single unit per the whole project so that when you invoke `jreleaserRelease` every one contributes to a single goal. This sounds like the root (or another single module) should orchestrate everything in which case the orchestrator must have JReleaser tasks registered. If the root uses just `jreleaser-base` then we know that's not going to happen.

**Reply by gotson** [reply]
By default the execution of a Gradle task at the parent level would propagate to child projects.

In the spirit of https://github.com/jreleaser/jreleaser/issues/1231 we could imagine that triggering `jreleaserFullRelease` would trickle down from root to leaf projects, each doing what it needs.

But I understand that the current model doesn't allow that.

**Reply by aalmiray** [reply]
Indeed. Mapping JReleaser's model, context, and workflow steps to native Gradle constructs means sidelining the workflow engine and re-implement it with Gradle APIs. We already have a Gradle DSL model which leads to duplication. Having a Gradle based engine would mean more duplication 😞
